util/topsql/reporter/reporter.go

```diff
@@ -222,6 +242,47 @@ func (tsr *RemoteTopSQLReporter) RegisterPlan(planDigest []byte, normalizedBinar
 	}
 }
 
+var _ DataSinkRegisterer = &RemoteTopSQLReporter{}
+
+// Register implements DataSinkRegisterer interface.
+func (tsr *RemoteTopSQLReporter) Register(dataSink DataSink) error {
+	tsr.dataSinks.mu.Lock()
+	defer tsr.dataSinks.mu.Unlock()
+
+	select {
+	case <-tsr.ctx.Done():
+		return errors.New("reporter is closed")
+	default:
+		if len(tsr.dataSinks.m) >= 10 {
+			return errors.New("too many datasinks")
+		}
+
+		tsr.dataSinks.m[dataSink] = struct{}{}
+
+		if len(tsr.dataSinks.m) > 0 {
+			variable.TopSQLVariable.Enable.Store(true)
+		}
+
+		return nil
+	}
+}
+
+// Deregister implements DataSinkRegisterer interface.
+func (tsr *RemoteTopSQLReporter) Deregister(dataSink DataSink) {
+	tsr.dataSinks.mu.Lock()
+	defer tsr.dataSinks.mu.Unlock()
+
+	select {
+	case <-tsr.ctx.Done():
+	default:
+		delete(tsr.dataSinks.m, dataSink)
+
+		if len(tsr.dataSinks.m) == 0 {
+			variable.TopSQLVariable.Enable.Store(false)
+		}
```

Comment:
Why dataSink doesn't need to call `CloseDataSink`? In this case the reporter is also dropping it.

Reply:
Have changed `CloseDataSink` to `OnDeregisterFromReporter`. Clearer name answers this question.